### PR TITLE
Introduce and enable lit tests via CMake

### PR DESCRIPTION
**Pull request by marbre**
This introduces support for lit tests:
* Implements `iree_glob_lit_tests()`
* Adds CMakeLists.txt to the test subdirectories
* Enables the test subdirectories in the upper tree CMakeLists.txt files

At the moment, this only enables lit tests relying on `FileCheck` and `iree-opt`:
* Since the Bazel function `iree_setup_lit_package()` is skipped, there is no setup phase.
* In consequence `iree_glob_lit_tests()` directly depends on `FileCheck` and `iree-opt`.
* `iree-translate` can be added as an additional *hard* dependency to enable further tests without introducing the need to implement a setup phase.

### CODE REVIEW

File: build_tools/cmake/iree_glob_lit_tests.cmake [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  iree_package_name(_PACKAGE_NAME)
+  file(GLOB_RECURSE _TEST_FILES *.mlir)
+
+  foreach(_TEST_FILE ${_TEST_FILES})
+    get_filename_component(_TEST_FILE_LOCATION ${_TEST_FILE} DIRECTORY)
+    get_filename_component(_TEST_NAME ${_TEST_FILE} NAME_WE)
+    set(_NAME "${_PACKAGE_NAME}_${_TEST_NAME}")
+
+    add_test(NAME ${_NAME} COMMAND ${CMAKE_SOURCE_DIR}/iree/tools/run_lit.sh ${_TEST_FILE} ${CMAKE_SOURCE_DIR}/iree/tools/IreeFileCheck.sh WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/iree/tools)
+    set_tests_properties(${_NAME} PROPERTIES DEPENDS iree_tools_iree-opt LLVMFileCheck IreeFileCheck)
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_glob_lit_tests.cmake]: If IreeFileCheck already depends on LLVMFileCheck, is it necessary to list the dependency here as well? It seems like it shouldn't be.

Comment (marbre) [reply build_tools/cmake/iree_glob_lit_tests.cmake]: Shoudn't be.

**Comment by GMNGeoffrey** [comment]
> > Thanks! FYI I kind of hate iree_glob_lit_tests and have been thinking about changing it. It violates all kinds of good practices around explicitly declaring your dependencies. I've been meaning to argue with Stella about it :-P I think it wouldn't be that hard in bazel to swap it out for something with an explicit glob and some explicit deps (but still making the lit test runner part of the rule). I think it also wouldn't be necessary to hardcode deps like FileCheck and iree-opt. What would that look like on the cmake side? It seems like that would work fine?
> 
> Depending on the way you change the Bazel side, this will also work fine with CMake. However, I am not sure how an explicit glob or rather the corresponding Bazel rule should look like. In my mind, glob is never 100% explicit. Anyway, I am open minded for a different realization.

Ah yeah that was a weird phrasing. I mean make what it's globbing explicit in the rule, like `srcs = glob(["*.mlir"])`

**Comment by marbre** [comment]
> Ah yeah that was a weird phrasing. I mean make what it's globbing explicit in the rule, like `srcs = glob(["*.mlir"])`

Ah, you mean to pass what kind of files to glob within the function call and not hard coding this in the rule itself and probably also adding deps directly to the function call?

**Comment by GMNGeoffrey** [comment]
> Ah, you mean to pass what kind of files to glob within the function call and not hard coding this in the rule itself and probably also adding deps directly to the function call?

Yes exactly. And I think there's not really a reason you need a separate setup rule. We should be able to gen those both as part of the rule, which I think would map more naturally to cmake. I'll take a look at doing that this week.

**Comment by marbre** [comment]
> Yes exactly. And I think there's not really a reason you need a separate setup rule. We should be able to gen those both as part of the rule, which I think would map more naturally to cmake. I'll take a look at doing that this week.

I think this should be straight forward with Bazel as well as with CMake. Just assign me an issue if you don't want to care for CMake by yourself :) I also have a clue howto adjust the Bazel configuration, but I think you also need to adjust it internal?

**Comment by GMNGeoffrey** [comment]
Hey it looks like the additional commit you made somehow broke the cmake build (sorry I should've requested another kokoro run before merging): https://source.cloud.google.com/results/invocations/504a14e0-4554-413b-84a8-f3c1eb091993

```
CMake Error at build_tools/cmake/iree_glob_lit_tests.cmake:35 (set_tests_properties):
  set_tests_properties called with incorrect number of arguments.
```

I'm not sure why that is exactly. It looks like you need a format like `prop val prop val` ( https://cmake.org/cmake/help/v3.12/command/set_tests_properties.html). My guess is that previously it was interpreting LLVMFileCheck as a property, so that worked (but probably not in the way you intended)

**Comment by GMNGeoffrey** [comment]
https://github.com/google/iree/pull/540 should fix this

**Comment by marbre** [comment]
> Hey it looks like the additional commit you made somehow broke the cmake build (sorry I should've requested another kokoro run before merging): https://source.cloud.google.com/results/invocations/504a14e0-4554-413b-84a8-f3c1eb091993
> 
> ```
> CMake Error at build_tools/cmake/iree_glob_lit_tests.cmake:35 (set_tests_properties):
>   set_tests_properties called with incorrect number of arguments.
> ```
> 
> I'm not sure why that is exactly. It looks like you need a format like `prop val prop val` ( https://cmake.org/cmake/help/v3.12/command/set_tests_properties.html). My guess is that previously it was interpreting LLVMFileCheck as a property, so that worked (but probably not in the way you intended)

I should  have tested locally prior to committing, sorry. Thanks for catching and fixing this.

**Pull request closed without merging**